### PR TITLE
feat(*): provide more `control` fields

### PR DESCRIPTION
**Pull request by oklopfer**
## Purpose

pull in more debian control fields

## Approach

adds the following:
- `enhances` (+ enhanced arrays)
- `recommends`  (+ enh arrs)
- `makeconflicts` (+ enh arrs)
- `checkconflicts` (+ enh arrs)
- `custom_fields` (NO enhanced arrays)
    - has a check to ensure field is not already used by pacstall, and is valid to debian format
        - no space
        - no numbers
        - first uppercase, rest lowercase
        - hyphen not trailing
        - hyphenation follows same case rules
    - array formatted such as:
```bash
 custom_fields=(
  "Example-Field: this, that, those"
  "Exampler: This one is a long description
that goes on multiple lines"
)
```
- breaks `conflicts` off of `replaces` ~~(run tests to ensure this is okay, might need to include `conflicts=("${replaces[@]}")` in packages)~~ can confirm this is all good

 

## Progress

- [x] do it
- [x] test it

## Checklist

- [x] I confirm that I have read the [contributing guidelines](https://github.com/pacstall/pacstall/blob/develop/CONTRIBUTING.md), and this pull request is abiding by all the clauses stated in the guideline.

### CODE REVIEW

**Comment by oklopfer** [comment]
all werks

File: misc/scripts/checks.sh [hunk 1]
```diff
@@ -356,6 +356,53 @@ function lint_relations() {
...
 }
 
+function lint_field_fmt() {
+    local infield="${1}"
+    # ensure no spaces or numbers, first letter is capital, + hyphen is not last char and also follows cap rule
+    if [[ "${infield}" =~ [[:space:]] ]]; then
+        return 1
+    elif [[ "${infield}" =~ [0-9] ]]; then
+        return 2
+    elif [[ ! "${infield}" =~ ^[A-Z][a-z]*(-[A-Z][a-z]*)*$ ]]; then
```
Comment (Elsie19) [inline misc/scripts/checks.sh:366]:
```suggestion
    elif [[ "${infield}" != "${infield^}" ]]; then
```

**Review by Elsie19 — CHANGES_REQUESTED** [review]
The `^` operator in variables uppercases the first character.

**Comment by Elsie19** [comment]
```patch
From daa685f1874ed8bab3fd6e7b4e479b1327045bb8 Mon Sep 17 00:00:00 2001
From: Elsie <elsie19@pm.me>
Date: Mon, 3 Jun 2024 19:23:48 -0500
Subject: [PATCH] Split up linting custom fields

---
 misc/scripts/checks.sh | 162 +++++++++++++++++++++++++----------------
 1 file changed, 101 insertions(+), 61 deletions(-)

diff --git a/misc/scripts/checks.sh b/misc/scripts/checks.sh
index 3f17f3f..9b614ee 100755
--- a/misc/scripts/checks.sh
+++ b/misc/scripts/checks.sh
@@ -171,42 +171,42 @@ function lint_source() {
     else
         for sarch in "${PACSTALL_KNOWN_ARCH[@]}" "${PACSTALL_KNOWN_DISTROS[@]}" "${source_distro_archs[@]}"; do
             local source_arch="source_${sarch}[@]" raw_carch_source="source_${TARCH}[@]" raw_distbase_source="source_${DISTRO%:*}[@]" \
-            raw_distver_source="source_${DISTRO#*:}[@]" raw_distbase_carch_source="source_${DISTRO%:*}_${TARCH}[@]" \
-            raw_distver_carch_source="source_${DISTRO#*:}_${TARCH}[@]" carch_source distbase_source distver_source distbase_carch_source distver_carch_source
+                raw_distver_source="source_${DISTRO#*:}[@]" raw_distbase_carch_source="source_${DISTRO%:*}_${TARCH}[@]" \
+                raw_distver_carch_source="source_${DISTRO#*:}_${TARCH}[@]" carch_source distbase_source distver_source distbase_carch_source distver_carch_source
             carch_source=("${!raw_carch_source}")
             distbase_source=("${!raw_distbase_source}")
             distver_source=("${!raw_distver_source}")
             distbase_carch_source=("${!raw_distbase_carch_source}")
             distver_carch_source=("${!raw_distver_carch_source}")
-            [[ ${sarch} != "${TARCH}"
-                && ${sarch} != "${DISTRO%:*}"
-                && ${sarch} != "${DISTRO#*:}"
-                && ${sarch} != "${DISTRO%:*}_${TARCH}"
-                && ${sarch} != "${DISTRO#*:}_${TARCH}"
-            ]] && if [[ -n ${!source_arch} ]]; then
-                test_source=()
-                if [[ -n ${source[0]} ]]; then
-                    { (( "${#source[@]}" <= 1
-                        && "${#carch_source[@]}" <= 1
-                        && "${#distbase_source[@]}" <= 1
-                        && "${#distver_source[@]}" <= 1
-                        && "${#distbase_carch_source[@]}" <= 1
-                        && "${#distver_carch_source[@]}" <= 1 )) \
-                        && [[ ${carch_source[0]} == "${source[0]}"
-                            || ${distbase_source[0]} == "${source[0]}"
-                            || ${distver_source[0]} == "${source[0]}"
-                            || ${distbase_carch_source[0]} == "${source[0]}"
-                            || ${distver_carch_source[0]} == "${source[0]}" ]]; } \
-                    || test_source+=("${source[@]}")
-                fi
-                test_source+=("${!source_arch}")
-                if [[ -n ${test_source[1]} ]]; then
-                    lint_source_deb_test "${test_source[@]}"
-                    if ((ret == 1)); then
-                        break
+            [[ ${sarch} != "${TARCH}" &&
+                ${sarch} != "${DISTRO%:*}" &&
+                ${sarch} != "${DISTRO#*:}" &&
+                ${sarch} != "${DISTRO%:*}_${TARCH}" &&
+                ${sarch} != "${DISTRO#*:}_${TARCH}" ]] \
+                && if [[ -n ${!source_arch} ]]; then
+                    test_source=()
+                    if [[ -n ${source[0]} ]]; then
+                        { (("${#source[@]}" <= 1 && \
+                            "${#carch_source[@]}" <= 1 && \
+                            "${#distbase_source[@]}" <= 1 && \
+                            "${#distver_source[@]}" <= 1 && \
+                            "${#distbase_carch_source[@]}" <= 1 && \
+                            "${#distver_carch_source[@]}" <= 1)) \
+                                && [[ ${carch_source[0]} == "${source[0]}" ||
+                                    ${distbase_source[0]} == "${source[0]}" ||
+                                    ${distver_source[0]} == "${source[0]}" ||
+                                    ${distbase_carch_source[0]} == "${source[0]}" ||
+                                    ${distver_carch_source[0]} == "${source[0]}" ]]; } \
+                            || test_source+=("${source[@]}")
+                    fi
+                    test_source+=("${!source_arch}")
+                    if [[ -n ${test_source[1]} ]]; then
+                        lint_source_deb_test "${test_source[@]}"
+                        if ((ret == 1)); then
+                            break
+                        fi
                     fi
                 fi
-            fi
         done
         if [[ -n ${source[1]} ]]; then
             lint_source_deb_test "${source[@]}"
@@ -356,15 +356,41 @@ function lint_relations() {
     return "${ret}"
 }

+function capital_checking() {
+    local str="${1}" i=0 c x z split_chars=()
+    while printf -v c "%s%n" "${str:i++:1}" x; do
+        ((x)) && split_chars+=("${c}") || break
+    done
+    for z in "${!split_chars[@]}"; do
+        if [[ ${split_chars[$z]} == '-' ]]; then
+            # Is the next letter a capital?
+            if [[ ${split_chars[z + 1]} != "${split_chars[z + 1]^}" ]]; then
+                return 1
+            fi
+        fi
+    done
+}
+
 function lint_field_fmt() {
     local infield="${1}"
-    # ensure no spaces or numbers, first letter is capital, + hyphen is not last char and also follows cap rule
-    if [[ "${infield}" =~ [[:space:]] ]]; then
+    # Ensure no spaces
+    if [[ ${infield} =~ [[:space:]] ]]; then
         return 1
-    elif [[ "${infield}" =~ [0-9] ]]; then
+    # Ensure no numbers
+    elif [[ ${infield} =~ [0-9] ]]; then
         return 2
-    elif [[ ! "${infield}" =~ ^[A-Z][a-z]*(-[A-Z][a-z]*)*$ ]]; then
+    # Ensure first letter is Capital
+    elif [[ ${infield} != "${infield^}" ]]; then
         return 3
+        # Ensure names with multiple words are hyphen separated, and don't end with a hyphen
+    elif [[ ${infield} != "${infield^}" ]]; then
+        return 4
+        # Ensure hyphenated field names are capitalized only on the first letter of each word
+    elif ! capital_checking "${infield}"; then
+        return 5
+        # Ensure last or first letter isn't a hyphen
+    elif [[ ${infield: -1} != '-' || ${infield:1} != '-' ]]; then
+        return 6
     else
         return 0
     fi
@@ -387,13 +413,28 @@ function lint_fields() {
                 fancy_message error "'${tlogvar}' is already used as a field in pacstall"
                 ret=1
             else
-			    lint_field_fmt "${tlogvar}"
+                lint_field_fmt "${tlogvar}"
                 case "$?" in
-                    1) fancy_message error "'${tlogvar}' field name cannot contain a space in 'custom_fields'"; ret=1 ;;
-                    2) fancy_message error "'${tlogvar}' field name cannot contain a number in 'custom_fields'"; ret=1 ;;
-                    3) fancy_message error "'${tlogvar}' field name must capitalize only the first letter in 'custom_fields'"
-                       fancy_message sub "Field names with multiple words should be hyphen separated, and cannot end with a hyphen"
-                       fancy_message sub "Hyphenated field names must capitalize only the first letter of each word"; ret=1 ;;
+                    1)
+                        fancy_message error "'${tlogvar}' field name cannot contain a space in 'custom_fields'"
+                        ret=1
+                        ;;
+                    2)
+                        fancy_message error "'${tlogvar}' field name cannot contain a number in 'custom_fields'"
+                        ret=1
+                        ;;
+                    3 | 5)
+                        fancy_message error "'${tlogvar}' field name must capitalize only the letter of each word in 'custom_fields'"
+                        ret=1
+                        ;;
+                    4)
+                        fancy_message error "'${tlogvar}' field name with multiple words should be hyphen separated in 'custom_fields'"
+                        ret=1
+                        ;;
+                    6)
+                        fancy_message error "'${tlogvar}' cannot start or end with a hyphen in 'custom_fields'"
+                        ret=1
+                        ;;
                 esac
             fi
             ((idx++))
@@ -402,7 +443,6 @@ function lint_fields() {
     fi
 }

-
 function lint_hash() {
     local ret=0 test_hash harch test_hashsum_type test_hashsum_style test_hash_arch test_hashsum_method test_hashsum_value \
         hash_distro_archs
@@ -433,33 +473,33 @@ function lint_hash() {
         test_hashsum_style="${test_hashsum_type}sums[*]"
         for harch in "${PACSTALL_KNOWN_ARCH[@]}" "${PACSTALL_KNOWN_DISTROS[@]}" "${hash_distro_archs[@]}"; do
             test_hash_arch="${test_hashsum_type}sums_${harch}[*]"
-            [[ ${harch} != "${TARCH}"
-                && ${harch} != "${DISTRO%:*}"
-                && ${harch} != "${DISTRO#*:}"
-                && ${harch} != "${DISTRO%:*}_${TARCH}"
-                && ${harch} != "${DISTRO#*:}_${TARCH}"
-            ]] && if [[ -n ${!test_hash_arch} ]]; then
-                if [[ -z ${!test_hashsum_style} && -z ${test_hash[*]} ]]; then
-                    if [[ -z ${test_hashsum_method} ]]; then
+            [[ ${harch} != "${TARCH}" &&
+                ${harch} != "${DISTRO%:*}" &&
+                ${harch} != "${DISTRO#*:}" &&
+                ${harch} != "${DISTRO%:*}_${TARCH}" &&
+                ${harch} != "${DISTRO#*:}_${TARCH}" ]] \
+                && if [[ -n ${!test_hash_arch} ]]; then
+                    if [[ -z ${!test_hashsum_style} && -z ${test_hash[*]} ]]; then
+                        if [[ -z ${test_hashsum_method} ]]; then
+                            # shellcheck disable=SC2206
+                            test_hash=(${!test_hash_arch})
+                            test_hashsum_method="${test_hashsum_type}"
+                        else
+                            fancy_message error "Only one checksum method can be provided for hashes"
+                            unset test_hash
+                            ret=1
+                            break
+                        fi
+                    elif [[ -n ${test_hashsum_method} && ${test_hashsum_method} == "${test_hashsum_type}" ]]; then
                         # shellcheck disable=SC2206
-                        test_hash=(${!test_hash_arch})
-                        test_hashsum_method="${test_hashsum_type}"
+                        test_hash+=(${!test_hash_arch})
                     else
                         fancy_message error "Only one checksum method can be provided for hashes"
                         unset test_hash
                         ret=1
                         break
                     fi
-                elif [[ -n ${test_hashsum_method} && ${test_hashsum_method} == "${test_hashsum_type}" ]]; then
-                    # shellcheck disable=SC2206
-                    test_hash+=(${!test_hash_arch})
-                else
-                    fancy_message error "Only one checksum method can be provided for hashes"
-                    unset test_hash
-                    ret=1
-                    break
                 fi
-            fi
         done
     done
     # shellcheck disable=SC2128
@@ -566,9 +606,9 @@ function lint_arch() {
                 ret=1
             else
                 for key in "${!AARCHS_MAP[@]}"; do
-                    if [[ "${el_arch}" == "${AARCHS_MAP[$key]}" ]]; then
+                    if [[ ${el_arch} == "${AARCHS_MAP[$key]}" ]]; then
                         has_aarch=true
-                    elif [[ "${el_arch}" == "${key}" ]]; then
+                    elif [[ ${el_arch} == "${key}" ]]; then
                         has_carch=true
                     fi
                 done
--
2.43.0

```

Apply that patch.

**Review by Elsie19 — APPROVED** [review]

**Pull request merged**